src/auxiliary/mpi_arrays.jl

```diff
@@ -0,0 +1,332 @@
+
+# TODO: MPI. Keep this module inside Trixi or move it to another repo as
+#            external dependency with simple test suite and documentation?
+module TrixiMPIArrays
+
+using ArrayInterface: ArrayInterface
+using MPI: MPI
+
+import ..Trixi: mpi_comm, mpi_rank
+
+export TrixiMPIArray
+
+
+"""
+    TrixiMPIArray{T, N} <: AbstractArray{T, N}
+
+A thin wrapper of arrays distributed via MPI used in Trixi.jl. The idea is that
+these arrays behave as much as possible as plain arrays would in an SPMD-style
+distributed MPI setting with exception of reductions, which are performed
+globally. This allows to use these arrays in ODE solvers such as the ones from
+OrdinaryDiffEq.jl, since vector space operations, broadcasting, and reductions
+are the only operations required for explicit time integration methods with
+fixed step sizes or adaptive step sizes based on CFL or error estimates.
+
+!!! warning "Experimental code"
+    This code is experimental and may be changed or removed in any future release.
+"""
+struct TrixiMPIArray{T, N, Parent<:AbstractArray{T, N}} <: AbstractArray{T, N}
+  u_local::Parent
+  mpi_comm::MPI.Comm
+  mpi_rank::Int
+  # TODO: MPI. Shall we also include something like the following fields
+  #       and remove them from the global state? Do we ever need something
+  #       from the global MPI state without having a state vector `u`? Does
+  #       including these fields here have a performance impact since it
+  #       increases the size of these arrays?
+  # mpi_size::Int
+  # mpi_isroot::Bool
+  # mpi_isparallel::Bool
```

Comment:
I think these are orthogonal issues. In a reasonably complex MPI simulation, it makes sense to have multiple MPI communicators. Thus a global communicator becomes - at some point - unsuitable. OTOH, all three fields can be computed on the fly from the `mpi_comm` - but so can `mpi_rank` - thus my suggestion woulud be to keep `mpi_size` (if it is ever needed) and ignore `mpi_isroot` and `mpi_isparallel` for now.